mbed_targets/_internal/target_attribute_hierarchy_parsers/accumulating_attribute_parser.py

```diff
@@ -0,0 +1,150 @@
+
+
+def _calculate_attribute_elements(
+    attribute_name: str, starting_state: Dict[str, Any], applicable_accumulation_order: List[dict]
+) -> Dict[str, Any]:
+    """Adds and removes elements for an attribute based on the definitions encountered in the hierarchy.
+
+    Args:
+        attribute_name: name of the attribute to update
+        starting_state: the list of elements that defines the starting state of the attribute
+        applicable_accumulation_order: the targets in the inheritance tree that can contain add and remove modifiers
+
+    Returns:
+        The elements calculated for an attribute according to the target's inheritance
+    """
+    accumulator = starting_state
+    for target in reversed(applicable_accumulation_order):
+        if f"{attribute_name}_add" in target.keys():
+            to_add = target[f"{attribute_name}_add"]
+            _add_attribute_element(accumulator, attribute_name, to_add)
+        if f"{attribute_name}_remove" in target.keys():
+            to_remove = target[f"{attribute_name}_remove"]
+            _remove_attribute_element(accumulator, attribute_name, to_remove)
+    return accumulator
+
+
+def _calculate_attribute_for_target(
+    attribute_name: str, target: Dict[str, Any], accumulation_order: List[dict]
+) -> Dict[str, Any]:
+    """Finds a single accumulated attribute for a target from its list of ancestors.
+
+    Args:
+        attribute_name: the name of the accumulating attribute
+        target: the target we are collecting data for
+        accumulation_order: the full inheritance hierarchy
+
+    Returns:
+        A dictionary representation of a single accumulating attribute for that target
+    """
+    starting_state = {attribute_name: target[attribute_name]}
+    # Reduces the order list to only the targets in the hierarchy between the starting state and the target itself
+    applicable_accumulation_order = accumulation_order[: accumulation_order.index(target)]
+    return _calculate_attribute_elements(attribute_name, starting_state, applicable_accumulation_order)
+
+
+def _determine_accumulated_attributes(accumulation_order: List[dict]) -> Dict[str, Any]:
+    """Finds all the accumulated attributes for a target from its list of ancestors.
+
+    Iterates through the order of inheritance (accumulation order) to find the nearest definition
+    of an attribute, then retraces backwards calculating additions and deletions that modify it.
+
+    Args:
+        accumulation_order: the order in which the targets inherit from the target itself to its highest ancestor
+
+    Returns:
+        A dictionary containing all the accumulating attributes for a target
+    """
+    accumulated_attributes = {}
+
+    target_attribute_pairs = itertools.product(accumulation_order, ACCUMULATING_ATTRIBUTES)
+    filtered_pairs = ((target, attribute) for target, attribute in target_attribute_pairs if attribute in target.keys())
+    for target, attribute_name in filtered_pairs:
```

Comment:
I think this is a bit hard to read, could you not do something like:
```
for tgt, attr in itertools.product(accumulation_order, ACCUMULATING_ATTRIBUTES):
    if attr in tgt.keys():
        accumulated_sttrs.update(_calc...)
```
Or does that not work?

Reply:
Yeah that is much better, thanks - Ill get there in the end 😂

Reply:
3beb9db